pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,380 @@
+        """Int. Conversion of ``control_values`` to an integer."""
+        return sum(2**i for i, val in enumerate(reversed(self.control_values)) if val)
+
+    # Properties on the wires ##########################
+
+    @property
+    def control_wires(self):
+        """The control wires."""
+        return self.hyperparameters["control_wires"]
+
+    @property
+    def target_wires(self):
+        """The wires of the target operator."""
+        return self.base.wires
+
+    @property
+    def work_wires(self):
+        """Additional wires that can be used in the decomposition. Not modified by the operation."""
+        return self.hyperparameters["work_wires"]
+
+    @property
+    def wires(self):
+        return self.control_wires + self.base.wires + self.work_wires
+
+    # pylint: disable=protected-access
+    @property
+    def _wires(self):
+        return self.wires
+
+    # pylint: disable=protected-access
+    @_wires.setter
+    def _wires(self, new_wires):
+        new_wires = new_wires if isinstance(new_wires, Wires) else Wires(new_wires)
+
+        num_control = len(self.control_wires)
+        num_base = len(self.base.wires)
+        num_control_and_base = num_control + num_base
+
+        assert num_control_and_base <= len(new_wires), (
+            f"{self.name} needs at least {num_control_and_base} wires."
+            f" {len(new_wires)} provided."
+        )
+
+        self.hyperparameters["control_wires"] = new_wires[:num_control]
+
+        self.base._wires = new_wires[num_control:num_control_and_base]
+
+        if len(new_wires) > num_control_and_base:
+            self.hyperparameters["work_wires"] = new_wires[num_control_and_base:]
+        else:
+            self.hyperparameters["work_wires"] = Wires([])
+
+    # Methods ##########################################
+
+    def label(self, decimals=None, base_label=None, cache=None):
+        return self.base.label(decimals=decimals, base_label=base_label, cache=cache)
+
+    def matrix(self, wire_order=None):
+
+        base_matrix = self.base.matrix()
+        interface = qmlmath.get_interface(base_matrix)
+
+        num_target_states = 2 ** len(self.target_wires)
+        num_control_states = 2 ** len(self.control_wires)
+        total_matrix_size = num_control_states * num_target_states
+
+        padding_left = self._control_int * num_target_states
+        padding_right = total_matrix_size - padding_left - num_target_states
+
+        left_pad = qmlmath.cast_like(qmlmath.eye(padding_left, like=interface), 1j)
+        right_pad = qmlmath.cast_like(qmlmath.eye(padding_right, like=interface), 1j)
+
+        canonical_matrix = qmlmath.block_diag([left_pad, base_matrix, right_pad])
+
+        if wire_order is None or self.wires == Wires(wire_order):
+            return canonical_matrix
+
+        active_wires = self.control_wires + self.target_wires
+        return operation.expand_matrix(canonical_matrix, wires=active_wires, wire_order=wire_order)
```

Comment:
Just a clarification here, so `self.wires` will have length `len(control_wires) + len(base.wires) + len(work_wires)` but the matrix here will only be of size  ` 2**(len(base.wires) + len(control_wires)) x 2**(len(base.wires) + len(control_wires))` correct?